lib/di/interceptor/app_interceptor.dart

```diff
@@ -1,26 +1,86 @@
+import 'dart:io';
+
 import 'package:dio/dio.dart';
+import 'package:flutter_survey/di/di.dart';
 import 'package:flutter_survey/local/shared_preference_helper.dart';
+import 'package:flutter_survey/models/oauth_token.dart';
+import 'package:flutter_survey/usecase/base/base_use_case.dart';
+import 'package:flutter_survey/usecase/refresh_token_use_case.dart';
 
-const String HEADER_AUTHORIZATION = 'Authorization';
+const String _headerAuthorization = 'Authorization';
 
-class AppInterceptor extends Interceptor {
-  bool _requireAuthenticate;
-  SharedPreferencesHelper _sharedPreferencesHelper;
+class AppInterceptor extends QueuedInterceptor {
+  final bool _requireAuthenticate;
+  final SharedPreferencesHelper _sharedPreferencesHelper;
+  final Dio _dio;
 
   AppInterceptor(
     this._requireAuthenticate,
     this._sharedPreferencesHelper,
+    this._dio,
   );
 
   @override
-  Future onRequest(
+  void onRequest(
       RequestOptions options, RequestInterceptorHandler handler) async {
     if (_requireAuthenticate) {
-      String accessToken = await _sharedPreferencesHelper.getAccessToken();
-      String tokenType = await _sharedPreferencesHelper.getTokenType();
-      options.headers
-          .putIfAbsent(HEADER_AUTHORIZATION, () => "$tokenType $accessToken");
+      final token = await _buildToken();
+      options.headers.putIfAbsent(_headerAuthorization, () => token);
+    }
+    super.onRequest(options, handler);
+  }
+
+  @override
+  void onError(DioError err, ErrorInterceptorHandler handler) {
+    final statusCode = err.response?.statusCode;
+    if ((statusCode == HttpStatus.forbidden ||
+            statusCode == HttpStatus.unauthorized) &&
+        _requireAuthenticate) {
+      _doRefreshToken(err, handler);
+    } else {
+      handler.next(err);
     }
-    return super.onRequest(options, handler);
+  }
+
+  Future<void> _doRefreshToken(
+    DioError err,
+    ErrorInterceptorHandler handler,
+  ) async {
+    try {
+      final refreshTokenUseCase = getIt<RefreshTokenUseCase>();
+      final result = await refreshTokenUseCase.call();
```

Comment:
just a comment, we could improve the logic here to avoid multiple API failed leading to multiple refresh token API calls at the same time by locking all failed requests until the refresh token API is done.

Reply:
I will make a test to investigate this issue later, thanks for reminding 🏃